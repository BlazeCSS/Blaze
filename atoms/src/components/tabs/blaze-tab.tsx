import { Component, Prop } from '@stencil/core';

@Component({
  tag: 'blaze-tab'
})
export class Tab {
  @Prop() header: string;
  @Prop() disabled: boolean;
  @Prop() open: boolean;

  render() {
    const openClass = this.open ? 'c-tabs__tab--active' : '';

    return (
      <div class={`c-tabs__tab ${openClass}`}>
        <slot />
      </div>
    );
  }
}
