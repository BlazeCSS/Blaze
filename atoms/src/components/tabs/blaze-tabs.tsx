import { Component, State, Prop, Method, Element } from '@stencil/core';

@Component({
  tag: 'blaze-tabs'
})
export class Tabs {
  @Element() elem: HTMLElement;
  @Prop() type: string;
  @State() tabs: any[];

  componentWillLoad() {
    this.tabs = Array.from(this.elem.querySelectorAll('blaze-tab'));
  }

  @Method()
  currentTab() {
    return this.tabs.findIndex(tab => tab.open)
  }

  @Method()
  openTab(tabIndex: number) {
    if (!this.tabs[tabIndex].disabled) {

      this.tabs = this.tabs.map(tab => {
        tab.open = false;
        return tab;
      });
      this.tabs[tabIndex].open = true;
    }
  }

  render() {
    const typeClass = this.type ? `c-tabs--${this.type}` : '';

    return (
      <div class={`c-tabs ${typeClass}`}>
        <div class="c-tabs__nav">
          <div class="c-tabs__headings">
            {
              this.tabs.map((tab, i: number) => {
                const disabledClass = tab.disabled ? 'c-tab-heading--disabled' : '';
                const openClass = tab.open ? 'c-tab-heading--active' : '';

                return (
                  <div class={`c-tab-heading ${disabledClass} ${openClass}`} onClick={() => this.openTab(i)}>
                    {tab.header}
                  </div>
                )
              })
            }
          </div>
        </div>
        <slot />
      </div>
    );
  }
}
