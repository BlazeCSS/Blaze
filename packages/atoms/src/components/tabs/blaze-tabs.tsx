import { Component, State, Method, Element } from '@stencil/core';

@Component({
  tag: 'blaze-tabs'
})
export class Tabs {
  @Element() elem: HTMLElement;
  @State() tabs: any[];

  componentWillLoad() {
    this.tabs = Array.from(this.elem.querySelectorAll('blaze-tab'));
  }

  @Method()
  currentTab() {
    return this.tabs.findIndex(tab => tab.open)
  }

  @Method()
  openTab(tabIndex: number) {
    if (!this.tabs[tabIndex].disabled) {

      this.tabs = this.tabs.map(tab => {
        tab.open = false;
        return tab;
      });
      this.tabs[tabIndex].open = true;
    }
  }

  render() {
    return (
      <div class="c-tabs">
        <div role="tablist" class="c-tabs">
          <div class="c-tabs__nav">
            <div class="c-tabs__headings">
              {
                this.tabs.map((tab, i: number) => {
                  const openClass = tab.open ? 'c-tab-heading--active' : '';
                  const typeClass = tab.type ? `c-tab-heading--${tab.type}` : '';

                  return (
                    <button role="tab" disabled={tab.disabled} class={`c-tab-heading ${typeClass} ${openClass}`} onClick={() => this.openTab(i)}>
                      {tab.header}
                    </button>
                  )
                })
              }
            </div>
          </div>
          <slot />
        </div>
      </div>
    );
  }
}
