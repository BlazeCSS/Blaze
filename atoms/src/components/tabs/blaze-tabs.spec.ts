import { flush, render } from '@stencil/core/testing';
import { Tabs } from './blaze-tabs';
import { Tab } from './blaze-tab';

describe('Tabs', () => {
  it('should build', () => {
    expect(new Tabs()).toBeTruthy();
  });

  let element;

  const snapIt = (name, html) => {
    it(name, async () => {
      element = await render({
        components: [Tabs],
        html
      });
      await flush(element);

      expect(element).toMatchSnapshot();
    });
  };

  snapIt(
    'renders correctly with no props',
    `<blaze-tabs>
      <blaze-tab>This is tab one</blaze-tab>
      <blaze-tab>This is tab two</blaze-tab>
    </blaze-tabs>`
  );

  snapIt(
    'renders correctly with props set',
    `<blaze-tabs type="brand">
      <blaze-tab header="Tab one" open>This is tab one</blaze-tab>
      <blaze-tab header="Tab two" disabled>This is tab two</blaze-tab>
    </blaze-tabs>`
  );

  it('returns current tab', async () => {
    element = await render({
      components: [Tabs, Tab],
      html: `<blaze-tabs type="brand">
      <blaze-tab header="Tab one">This is tab one</blaze-tab>
      <blaze-tab header="Tab two" open>This is tab two</blaze-tab>
    </blaze-tabs>`
    });
    await flush(element);

    expect(element.currentTab()).toEqual(1);
  });

  it('switches tabs', async () => {
    element = await render({
      components: [Tabs, Tab],
      html: `<blaze-tabs type="brand">
      <blaze-tab header="Tab one" open>This is tab one</blaze-tab>
      <blaze-tab header="Tab two">This is tab two</blaze-tab>
    </blaze-tabs>`
    });
    await flush(element);
    expect(element.currentTab()).toEqual(0);

    element.openTab(1);
    await flush(element);

    expect(element.currentTab()).toEqual(1);
  });

  it('switches tabs by clicking on them', async () => {
    element = await render({
      components: [Tabs, Tab],
      html: `<blaze-tabs type="brand">
      <blaze-tab header="Tab one">This is tab one</blaze-tab>
      <blaze-tab header="Tab two" open>This is tab two</blaze-tab>
    </blaze-tabs>`
    });
    await flush(element);
    expect(element.currentTab()).toEqual(1);

    element.querySelector('.c-tab-heading').click();
    await flush(element);

    expect(element.currentTab()).toEqual(0);
  });
});
