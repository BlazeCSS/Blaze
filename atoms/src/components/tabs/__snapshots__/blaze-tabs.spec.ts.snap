// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders correctly with no props 1`] = `
<blaze-tabs
  class="hydrated"
>
  <div
    class="c-tabs"
  >
    <div
      class="c-tabs__nav"
    >
      <div
        class="c-tabs__headings"
      >
        <div
          class="c-tab-heading"
        >
          
        </div>
        <div
          class="c-tab-heading"
        >
          
        </div>
      </div>
    </div>
    <!---->
    
      
    <blaze-tab>
      This is tab one
    </blaze-tab>
    
      
    <blaze-tab>
      This is tab two
    </blaze-tab>
    
    
  </div>
</blaze-tabs>
`;

exports[`Tabs renders correctly with props set 1`] = `
<blaze-tabs
  class="hydrated"
  type="brand"
>
  <div
    class="c-tabs c-tabs--brand"
  >
    <div
      class="c-tabs__nav"
    >
      <div
        class="c-tabs__headings"
      >
        <div
          class="c-tab-heading"
        >
          
        </div>
        <div
          class="c-tab-heading"
        >
          
        </div>
      </div>
    </div>
    <!---->
    
      
    <blaze-tab
      header="Tab one"
      open=""
    >
      This is tab one
    </blaze-tab>
    
      
    <blaze-tab
      disabled=""
      header="Tab two"
    >
      This is tab two
    </blaze-tab>
    
    
  </div>
</blaze-tabs>
`;
